util/uri/src/fog.rs

```diff
@@ -0,0 +1,612 @@
+/// and is used when interacting with fog report server.
+/// It contains additional query parameters including a possible shard id,
+/// and a signature formed using the fog user's private keys over the
+/// fog root authority public key, used to authenticate reports from fog.
+#[derive(Debug, Hash, Ord, PartialOrd, Eq, PartialEq, Clone)]
+pub struct FogScheme {}
+
+impl UriScheme for FogScheme {
+    /// The part before the '://' of a URL.
+    const SCHEME_SECURE: &'static str = "fog";
+    const SCHEME_INSECURE: &'static str = "insecure-fog";
+
+    /// Default port numbers
+    const DEFAULT_SECURE_PORT: u16 = 443;
+    const DEFAULT_INSECURE_PORT: u16 = 3229;
+}
+
+/// Fog View Uri Scheme
+#[derive(Debug, Hash, Ord, PartialOrd, Eq, PartialEq, Clone)]
+pub struct FogViewScheme {}
+
+impl UriScheme for FogViewScheme {
+    /// The part before the '://' of a URL.
+    const SCHEME_SECURE: &'static str = "fog-view";
+    const SCHEME_INSECURE: &'static str = "insecure-fog-view";
+
+    /// Default port numbers
+    const DEFAULT_SECURE_PORT: u16 = 443;
+    const DEFAULT_INSECURE_PORT: u16 = 3225;
+}
+
+/// Fog Ledger Uri Scheme
+#[derive(Debug, Hash, Ord, PartialOrd, Eq, PartialEq, Clone)]
+pub struct FogLedgerScheme {}
+
+impl UriScheme for FogLedgerScheme {
+    /// The part before the '://' of a URL.
+    const SCHEME_SECURE: &'static str = "fog-ledger";
+    const SCHEME_INSECURE: &'static str = "insecure-fog-ledger";
+
+    /// Default port numbers
+    const DEFAULT_SECURE_PORT: u16 = 443;
+    const DEFAULT_INSECURE_PORT: u16 = 3223;
+}
+
+// Fog URI is the location of the fog service as a whole, fog view and fog ledger uri's
+// can be derived from it.
+// fog_uri.addr() itself is the location of the fog report server, and get_fog_report_label()
+// can be used to find the desired report in the response from fog report server.
+pub type FogUri = Uri<FogScheme>;
+pub type FogViewUri = Uri<FogViewScheme>;
+pub type FogLedgerUri = Uri<FogLedgerScheme>;
+
+// Extra API for FogUri
+pub trait FogUriApi {
+    /// Get the shard-id, if any, that is part of this Fog Uri.
+    fn get_shard_id(&self) -> Option<String>;
+
+    /// Get the user's signature over the fog authority's fingerprint, if any, that is part of this Fog Uri.
+    fn get_fog_authority_fingerprint_signature(&self) -> Option<String>;
+
+    /// Get the string label that Alice expects to find in report records from the report server when sending
+    /// to Bob's public address which has this fog uri.
+    ///
+    /// This is created by appending the shard id to the uri, and dropping any extra query parameters.
+    fn get_fog_report_label(&self) -> Result<String, UriParseError>;
+
+    /// Helper for implementing conversions from FogUri to other schemes
+    /// This appends the shard id to the url, and drops the fog authority fingerprint signature.
+    /// But preseves any tls / ca-bundle information etc.
+    fn append_shard_id_and_change_scheme<Scheme: UriScheme>(
```

Comment:
I was trying to avoid duplicating the code, but it's only being duplicated twice I guess. Or I can use a private fn I guess instead of sticking it into the trait.